File: examples/cpp/nlp/bert_example.cpp
/**
 * @file bert_example.cpp
 * @brief Example demonstrating BERT for text classification
 *
 * This example shows how to:
 * 1. Create and configure a BERT model for sentiment analysis
 * 2. Tokenize and prepare input text
 * 3. Train the model on sentiment classification
 * 4. Evaluate and make predictions
 * 5. Load pretrained weights (when available)
 */

#include <iostream>
#include <vector>
#include <string>
#include <memory>
#include <random>
#include <algorithm>
#include <iomanip>
#include <cstring>
#include "tenzor/models/bert.hpp"
#include "tenzor/core/tensor.hpp"
#include "tenzor/autograd/variable.hpp"
#include "tenzor/nn/optim/adam.hpp"
#include "tenzor/ops/math.hpp"
#include "tenzor/ops/reduction.hpp"
#include "tenzor/nn/activations/activations.hpp"

using namespace tenzor;
using namespace tenzor::models;
using namespace tenzor::nn;

// ============================================================================
// Simple Tokenizer (In practice, use a proper BPE tokenizer like Hugging Face)
// ============================================================================

class SimpleTokenizer {
public:
    SimpleTokenizer(int64_t vocab_size) : vocab_size_(vocab_size) {
        // Special tokens
        cls_token_id_ = 101;    // [CLS]
        sep_token_id_ = 102;    // [SEP]
        pad_token_id_ = 0;      // [PAD]
        unk_token_id_ = 100;    // [UNK]
    }

    /**
     * @brief Tokenize text into token IDs
     *
     * This is a very simplified tokenizer. In practice, use:
     * - Hugging Face tokenizers
     * - SentencePiece
     * - Custom BPE/WordPiece implementations
     */
    auto tokenize(const std::string& text, int64_t max_length) -> std::vector<int64_t> {
        std::vector<int64_t> token_ids;

        // Add [CLS] token
        token_ids.push_back(cls_token_id_);

        // Simple word splitting (replace with proper tokenizer)
        std::istringstream iss(text);
        std::string word;
        while (iss >> word && token_ids.size() < max_length - 1) {
            // Hash word to get token ID (very simplified)
            int64_t token_id = std::hash<std::string>{}(word) % (vocab_size_ - 200) + 200;
            token_ids.push_back(token_id);
        }

        // Add [SEP] token
        token_ids.push_back(sep_token_id_);

        // Pad to max_length
        while (token_ids.size() < max_length) {
            token_ids.push_back(pad_token_id_);
        }

        return token_ids;
    }

    /**
     * @brief Create attention mask (1 for real tokens, 0 for padding)
     */
    auto create_attention_mask(const std::vector<int64_t>& token_ids) -> std::vector<float> {
        std::vector<float> mask;
        for (auto token_id : token_ids) {
            mask.push_back(token_id != pad_token_id_ ? 1.0f : 0.0f);
        }
        return mask;
    }

private:
    int64_t vocab_size_;
    int64_t cls_token_id_;
    int64_t sep_token_id_;
    int64_t pad_token_id_;
    int64_t unk_token_id_;
};

// ============================================================================
// Dataset
// ============================================================================

struct SentimentExample {
    std::string text;
    int64_t label;  // 0 = negative, 1 = positive
};

class SentimentDataset {
public:
    SentimentDataset() {
        // Sample sentiment data
        examples_ = {
            {"This movie was absolutely fantastic and I loved every moment", 1},
            {"Terrible waste of time, completely disappointed", 0},
            {"Great performance by the actors and excellent cinematography", 1},
            {"Boring and predictable plot with weak characters", 0},
            {"One of the best films I have ever seen", 1},
            {"Awful script and poor direction throughout", 0},
            {"Highly recommended for anyone who enjoys quality cinema", 1},
            {"Not worth watching, saved you some time", 0},
            {"Brilliant storytelling and emotional depth", 1},
            {"Poorly executed with numerous plot holes", 0},
            {"Outstanding cast and beautiful visuals", 1},
            {"Disappointing ending after a slow start", 0},
            {"Masterpiece of modern filmmaking", 1},
            {"Complete disaster from beginning to end", 0},
            {"Engaging and thought-provoking throughout", 1},
            {"Waste of money and time, avoid at all costs", 0},
        };
    }

    auto size() const -> size_t {
        return examples_.size();
    }

    auto operator[](size_t idx) const -> const SentimentExample& {
        return examples_[idx];
    }

    auto get_examples() const -> const std::vector<SentimentExample>& {
        return examples_;
    }

private:
    std::vector<SentimentExample> examples_;
};

// ============================================================================
// Training Functions
// ============================================================================

auto create_batch(const std::vector<SentimentExample>& examples,
                 const std::vector<size_t>& indices,
                 SimpleTokenizer& tokenizer,
                 int64_t max_length,
                 Device device)
    -> std::tuple<Variable, Tensor, Tensor> {

    int64_t batch_size = indices.size();

    // Tokenize all examples
    std::vector<std::vector<int64_t>> all_token_ids;
    std::vector<std::vector<float>> all_masks;
    std::vector<int64_t> all_labels;

    for (auto idx : indices) {
        auto token_ids = tokenizer.tokenize(examples[idx].text, max_length);
        auto mask = tokenizer.create_attention_mask(token_ids);

        all_token_ids.push_back(token_ids);
        all_masks.push_back(mask);
        all_labels.push_back(examples[idx].label);
    }

    // Create input tensor on CPU, fill data, then move to device
    Tensor input_ids_cpu({batch_size, max_length}, DType::Int64, Device::cpu());
    std::vector<int64_t> flat_ids;
    for (const auto& ids : all_token_ids) {
        flat_ids.insert(flat_ids.end(), ids.begin(), ids.end());
    }
    int64_t* input_ptr = const_cast<int64_t*>(input_ids_cpu.data<int64_t>());
    std::memcpy(input_ptr, flat_ids.data(), flat_ids.size() * sizeof(int64_t));
    Tensor input_ids = input_ids_cpu.to(device);

    // Create attention mask on CPU, fill data, then move to device
    Tensor attention_mask_cpu({batch_size, max_length}, DType::Float32, Device::cpu());
    std::vector<float> flat_masks;
    for (const auto& mask : all_masks) {
        flat_masks.insert(flat_masks.end(), mask.begin(), mask.end());
    }
    float* mask_ptr = const_cast<float*>(attention_mask_cpu.data<float>());
    std::memcpy(mask_ptr, flat_masks.data(), flat_masks.size() * sizeof(float));
    Tensor attention_mask = attention_mask_cpu.to(device);

    // Create labels on CPU, fill data, then move to device
    Tensor labels_cpu({batch_size}, DType::Int64, Device::cpu());
    int64_t* labels_ptr = const_cast<int64_t*>(labels_cpu.data<int64_t>());
    std::memcpy(labels_ptr, all_labels.data(), all_labels.size() * sizeof(int64_t));
    Tensor labels = labels_cpu.to(device);

    return {Variable(input_ids, true), attention_mask, labels};
}

auto train_epoch(BertForSequenceClassification& model,
                SentimentDataset& dataset,
                SimpleTokenizer& tokenizer,
                optim::Adam& optimizer,
                int64_t batch_size,
                int64_t max_length,
                Device device) -> double {

    model.train();
    double total_loss = 0.0;
    int num_batches = 0;

    // Create indices and shuffle
    std::vector<size_t> indices(dataset.size());
    std::iota(indices.begin(), indices.end(), 0);

    std::random_device rd;
    std::mt19937 g(rd());
    std::shuffle(indices.begin(), indices.end(), g);

    // Process batches
    for (size_t start = 0; start < indices.size(); start += batch_size) {
        size_t end = std::min(start + batch_size, indices.size());
        std::vector<size_t> batch_indices(indices.begin() + start, indices.begin() + end);

        // Create batch
        auto [input_ids, attention_mask, labels] = create_batch(
            dataset.get_examples(), batch_indices, tokenizer, max_length, device);

        // Forward pass
        auto logits = model.forward(input_ids, attention_mask, Variable{});

        // Compute loss (simple mean squared error for demonstration)
        // In practice, use proper cross-entropy loss
        Variable labels_var(labels, false);
        auto diff = logits - labels_var;
        auto squared_tensor = pow(diff.tensor(), 2.0f);
        auto mean_tensor = mean(squared_tensor);
        auto loss = Variable(mean_tensor, true);

        // Backward pass
        optimizer.zero_grad();
        loss.backward();
        optimizer.step();

        auto loss_cpu = loss.tensor().to(Device::cpu());
        float* loss_data = const_cast<float*>(loss_cpu.data<float>());
        total_loss += loss_data[0];
        num_batches++;
    }

    return total_loss / num_batches;
}

auto evaluate(BertForSequenceClassification& model,
             SentimentDataset& dataset,
             SimpleTokenizer& tokenizer,
             int64_t batch_size,
             int64_t max_length,
             Device device) -> double {

    model.eval();
    int correct = 0;
    int total = 0;

    std::vector<size_t> indices(dataset.size());
    std::iota(indices.begin(), indices.end(), 0);

    for (size_t start = 0; start < indices.size(); start += batch_size) {
        size_t end = std::min(start + batch_size, indices.size());
        std::vector<size_t> batch_indices(indices.begin() + start, indices.begin() + end);

        auto [input_ids, attention_mask, labels] = create_batch(
            dataset.get_examples(), batch_indices, tokenizer, max_length, device);

        // Forward pass
        auto logits = model.forward(input_ids, attention_mask, Variable{});

        // Get predictions - move to CPU for comparison
        auto predictions = argmax(logits.tensor(), 1).to(Device::cpu());
        auto labels_cpu = labels.to(Device::cpu());

        // Count correct predictions
        const int64_t* pred_data = predictions.data<int64_t>();
        const int64_t* label_data = labels_cpu.data<int64_t>();

        for (int64_t i = 0; i < predictions.shape()[0]; ++i) {
            if (pred_data[i] == label_data[i]) {
                correct++;
            }
            total++;
        }
    }

    return static_cast<double>(correct) / total;
}

// ============================================================================
// Main
// ============================================================================

int main(int argc, char** argv) {
    std::cout << "=== BERT Text Classification Example ===" << std::endl;
    std::cout << std::endl;

    // Configuration
    int64_t max_length = 32;
    int64_t batch_size = 4;
    int64_t num_epochs = 10;
    double learning_rate = 2e-5;
    Device device = Device::cpu();

    std::cout << "Configuration:" << std::endl;
    std::cout << "  Max sequence length: " << max_length << std::endl;
    std::cout << "  Batch size: " << batch_size << std::endl;
    std::cout << "  Number of epochs: " << num_epochs << std::endl;
    std::cout << "  Learning rate: " << learning_rate << std::endl;
    std::cout << "  Device: " << device.to_string() << std::endl;
    std::cout << std::endl;

    // Create small BERT configuration for quick training
    BertConfig config;
    config.vocab_size = 5000;
    config.hidden_size = 128;
    config.num_hidden_layers = 2;
    config.num_attention_heads = 4;
    config.intermediate_size = 512;
    config.max_position_embeddings = 64;
    config.hidden_dropout_prob = 0.1;
    config.attention_probs_dropout_prob = 0.1;

    std::cout << "BERT Configuration:" << std::endl;
    std::cout << "  Vocabulary size: " << config.vocab_size << std::endl;
    std::cout << "  Hidden size: " << config.hidden_size << std::endl;
    std::cout << "  Number of layers: " << config.num_hidden_layers << std::endl;
    std::cout << "  Attention heads: " << config.num_attention_heads << std::endl;
    std::cout << std::endl;

    // Create model
    int64_t num_labels = 2;  // Binary classification (negative/positive)
    auto model = BertForSequenceClassification(config, num_labels);
    model.to(device);

    std::cout << "Created BERT model for sequence classification" << std::endl;
    std::cout << "Number of labels: " << num_labels << std::endl;

    // Count parameters
    auto params = model.parameters();
    size_t total_params = 0;
    for (const auto& param : params) {
        size_t param_size = 1;
        for (auto dim : param->tensor().shape()) {
            param_size *= dim;
        }
        total_params += param_size;
    }
    std::cout << "Total parameters: " << total_params << std::endl;
    std::cout << std::endl;

    // Create optimizer
    auto optimizer = optim::Adam(params, learning_rate);

    // Create dataset and tokenizer
    auto dataset = SentimentDataset();
    auto tokenizer = SimpleTokenizer(config.vocab_size);

    std::cout << "Dataset size: " << dataset.size() << " examples" << std::endl;
    std::cout << std::endl;

    // Display sample data
    std::cout << "Sample training examples:" << std::endl;
    for (size_t i = 0; i < std::min(size_t(3), dataset.size()); ++i) {
        const auto& example = dataset[i];
        std::cout << "  [" << (example.label == 1 ? "POSITIVE" : "NEGATIVE") << "] "
                  << example.text << std::endl;
    }
    std::cout << std::endl;

    // Training loop
    std::cout << "Starting training..." << std::endl;
    std::cout << std::string(60, '-') << std::endl;

    for (int64_t epoch = 1; epoch <= num_epochs; ++epoch) {
        auto epoch_loss = train_epoch(model, dataset, tokenizer, optimizer,
                                     batch_size, max_length, device);

        auto accuracy = evaluate(model, dataset, tokenizer,
                                batch_size, max_length, device);

        std::cout << "Epoch " << std::setw(2) << epoch << "/" << num_epochs
                  << " | Loss: " << std::fixed << std::setprecision(4) << epoch_loss
                  << " | Accuracy: " << std::setprecision(2) << (accuracy * 100) << "%"
                  << std::endl;
    }

    std::cout << std::string(60, '-') << std::endl;
    std::cout << std::endl;

    // Final evaluation
    auto final_accuracy = evaluate(model, dataset, tokenizer,
                                   batch_size, max_length, device);
    std::cout << "Final accuracy: " << std::fixed << std::setprecision(2)
              << (final_accuracy * 100) << "%" << std::endl;
    std::cout << std::endl;

    // Make predictions on new examples
    std::cout << "Making predictions on new examples:" << std::endl;
    std::cout << std::string(60, '-') << std::endl;

    model.eval();

    std::vector<std::string> test_texts = {
        "This is an amazing and wonderful experience",
        "Absolutely terrible and horrible quality",
        "Pretty good overall with some nice moments",
        "Not great but not terrible either"
    };

    for (const auto& text : test_texts) {
        // Tokenize
        auto token_ids = tokenizer.tokenize(text, max_length);
        auto mask = tokenizer.create_attention_mask(token_ids);

        // Create tensors on CPU, fill, then move to device
        Tensor input_ids_cpu({1, max_length}, DType::Int64, Device::cpu());
        int64_t* input_ptr = const_cast<int64_t*>(input_ids_cpu.data<int64_t>());
        std::memcpy(input_ptr, token_ids.data(), token_ids.size() * sizeof(int64_t));
        Tensor input_ids = input_ids_cpu.to(device);

        Tensor attention_mask_cpu({1, max_length}, DType::Float32, Device::cpu());
        float* mask_ptr = const_cast<float*>(attention_mask_cpu.data<float>());
        std::memcpy(mask_ptr, mask.data(), mask.size() * sizeof(float));
        Tensor attention_mask = attention_mask_cpu.to(device);

        // Predict
        auto logits = model.forward(Variable(input_ids, false), attention_mask, Variable{});
        auto probs = softmax(logits, 1);

        auto pred_tensor = argmax(logits.tensor(), 1).to(Device::cpu());
        auto prob_cpu = probs.tensor().to(Device::cpu());

        int64_t* pred_data = const_cast<int64_t*>(pred_tensor.data<int64_t>());
        int64_t prediction = pred_data[0];
        float* prob_data = const_cast<float*>(prob_cpu.data<float>());
        float confidence = prob_data[prediction];

        std::cout << "Text: " << text << std::endl;
        std::cout << "  Prediction: " << (prediction == 1 ? "POSITIVE" : "NEGATIVE")
                  << " (confidence: " << std::fixed << std::setprecision(2)
                  << (confidence * 100) << "%)" << std::endl;
    }

    std::cout << std::string(60, '-') << std::endl;
    std::cout << std::endl;

    // Save model
    std::string model_path = "/tmp/bert_sentiment_model.pt";
    std::cout << "Saving model to: " << model_path << std::endl;
    model.save(model_path);
    std::cout << "Model saved successfully" << std::endl;
    std::cout << std::endl;

    // Demonstrate loading pretrained weights (when available)
    std::cout << "Note: To use pretrained BERT weights:" << std::endl;
    std::cout << "  1. Download model from Hugging Face Hub" << std::endl;
    std::cout << "  2. Place checkpoint in ~/.cache/tenzor/bert-base-uncased/" << std::endl;
    std::cout << "  3. Call ModelHub::load_pretrained_weights(model, \"bert-base-uncased\")" << std::endl;
    std::cout << std::endl;

    std::cout << "Example completed successfully!" << std::endl;

    return 0;
}

File: examples/python/nlp/bert_example.py
"""
BERT Text Classification Example

This comprehensive example demonstrates:
- BERT model architecture components
- Simple tokenization for demonstration
- Sentiment classification training
- Attention mask handling
- Position embeddings

Components used:
- Embedding, LayerNorm, MultiheadAttention
- Linear, Dropout, GELU activation
- CrossEntropyLoss, Adam optimizer
"""

import tenzor as tz
import numpy as np


# ============================================================================
# Simple Tokenizer (for demonstration - use HuggingFace in practice)
# ============================================================================

class SimpleTokenizer:
    """Simplified tokenizer for demonstration purposes"""

    def __init__(self, vocab_size=30522):
        self.vocab_size = vocab_size
        self.cls_token_id = 101   # [CLS]
        self.sep_token_id = 102   # [SEP]
        self.pad_token_id = 0     # [PAD]
        self.unk_token_id = 100   # [UNK]

    def tokenize(self, text, max_length=128):
        """Tokenize text into token IDs"""
        token_ids = [self.cls_token_id]

        # Simple word-based tokenization (hash to get token ID)
        words = text.lower().split()
        for word in words:
            if len(token_ids) >= max_length - 1:
                break
            # Hash word to get token ID
            token_id = hash(word) % (self.vocab_size - 200) + 200
            token_ids.append(abs(token_id))

        # Add [SEP] token
        token_ids.append(self.sep_token_id)

        # Pad to max_length
        while len(token_ids) < max_length:
            token_ids.append(self.pad_token_id)

        return token_ids[:max_length]

    def create_attention_mask(self, token_ids):
        """Create attention mask (1 for real tokens, 0 for padding)"""
        return [1.0 if tid != self.pad_token_id else 0.0 for tid in token_ids]


# ============================================================================
# Sentiment Dataset
# ============================================================================

class SentimentDataset:
    """Simple sentiment classification dataset"""

    def __init__(self):
        self.examples = [
            ("This movie was absolutely fantastic and I loved every moment", 1),
            ("Terrible waste of time, completely disappointed", 0),
            ("Great performance by the actors and excellent cinematography", 1),
            ("Boring and predictable plot with weak characters", 0),
            ("One of the best films I have ever seen", 1),
            ("Awful script and poor direction throughout", 0),
            ("Highly recommended for anyone who enjoys quality cinema", 1),
            ("Not worth watching, saved you some time", 0),
            ("Brilliant storytelling and emotional depth", 1),
            ("Poorly executed with numerous plot holes", 0),
            ("Outstanding cast and beautiful visuals", 1),
            ("Disappointing ending after a slow start", 0),
            ("Masterpiece of modern filmmaking", 1),
            ("Complete disaster from beginning to end", 0),
            ("Engaging and thought-provoking throughout", 1),
            ("Waste of money and time, avoid at all costs", 0),
        ]

    def __len__(self):
        return len(self.examples)

    def __getitem__(self, idx):
        return self.examples[idx]


# ============================================================================
# BERT-like Classifier (simplified)
# ============================================================================

class BERTClassifier:
    """
    Simplified BERT-style classifier for demonstration

    Architecture:
    - Token + Position Embeddings
    - Multiple Transformer Encoder layers
    - [CLS] token classification head
    """

    def __init__(self, vocab_size, hidden_size, num_heads, num_layers,
                 num_classes, max_seq_len=128, dropout=0.1):
        self.hidden_size = hidden_size
        self.num_layers = num_layers

        # Embeddings
        self.token_embedding = tz.nn.Embedding(vocab_size, hidden_size)
        self.position_embedding = tz.nn.Embedding(max_seq_len, hidden_size)
        self.embedding_norm = tz.nn.LayerNorm([hidden_size])
        self.embedding_dropout = tz.nn.Dropout(dropout)

        # Transformer encoder layers
        self.attention_layers = []
        self.attention_norms = []
        self.ffn_layers = []
        self.ffn_norms = []

        for i in range(num_layers):
            self.attention_layers.append(
                tz.nn.MultiheadAttention(hidden_size, num_heads, dropout=dropout)
            )
            self.attention_norms.append(tz.nn.LayerNorm([hidden_size]))

            # Feed-forward network
            self.ffn_layers.append({
                'fc1': tz.nn.Linear(hidden_size, hidden_size * 4),
                'fc2': tz.nn.Linear(hidden_size * 4, hidden_size),
                'dropout': tz.nn.Dropout(dropout)
            })
            self.ffn_norms.append(tz.nn.LayerNorm([hidden_size]))

        # Classification head
        self.classifier = tz.nn.Linear(hidden_size, num_classes)
        self.classifier_dropout = tz.nn.Dropout(dropout)

    def forward(self, input_ids, attention_mask=None):
        """
        Forward pass

        Args:
            input_ids: Token IDs [batch, seq_len]
            attention_mask: Attention mask [batch, seq_len] (optional)

        Returns:
            Logits for classification [batch, num_classes]
        """
        batch_size = input_ids.tensor().shape[0]
        seq_len = input_ids.tensor().shape[1]

        # Create position IDs
        positions = np.arange(seq_len, dtype=np.int64)
        positions = np.tile(positions, (batch_size, 1))
        position_ids = tz.Variable(tz.Tensor.from_numpy(positions), requires_grad=False)

        # Token embeddings + position embeddings
        token_emb = self.token_embedding(input_ids)
        pos_emb = self.position_embedding(position_ids)

        # Combine embeddings
        hidden = tz.Variable(
            token_emb.tensor() + pos_emb.tensor(),
            requires_grad=True
        )
        hidden = self.embedding_norm(hidden)
        hidden = self.embedding_dropout(hidden)

        # Transformer encoder layers
        for i in range(self.num_layers):
            # Self-attention with residual (single input for self-attention)
            attn_out = self.attention_layers[i](hidden)
            hidden = tz.Variable(hidden.tensor() + attn_out.tensor(), requires_grad=True)
            hidden = self.attention_norms[i](hidden)

            # Feed-forward with residual
            ffn = self.ffn_layers[i]
            ffn_out = ffn['fc1'](hidden)
            ffn_out = tz.nn.gelu(ffn_out)
            ffn_out = ffn['fc2'](ffn_out)
            ffn_out = ffn['dropout'](ffn_out)

            hidden = tz.Variable(hidden.tensor() + ffn_out.tensor(), requires_grad=True)
            hidden = self.ffn_norms[i](hidden)

        # Extract [CLS] token representation (first token)
        hidden_np = hidden.tensor().numpy()
        cls_hidden = hidden_np[:, 0, :]  # [batch, hidden_size]
        cls_var = tz.Variable(tz.Tensor.from_numpy(cls_hidden), requires_grad=True)

        # Classification
        cls_var = self.classifier_dropout(cls_var)
        logits = self.classifier(cls_var)

        return logits

    def parameters(self):
        """Get all trainable parameters"""
        params = []
        params.extend(self.token_embedding.parameters())
        params.extend(self.position_embedding.parameters())
        params.extend(self.embedding_norm.parameters())

        for i in range(self.num_layers):
            params.extend(self.attention_layers[i].parameters())
            params.extend(self.attention_norms[i].parameters())
            params.extend(self.ffn_layers[i]['fc1'].parameters())
            params.extend(self.ffn_layers[i]['fc2'].parameters())
            params.extend(self.ffn_norms[i].parameters())

        params.extend(self.classifier.parameters())
        return params

    def train(self):
        """Set to training mode"""
        self.embedding_dropout.train()
        self.classifier_dropout.train()
        for i in range(self.num_layers):
            self.ffn_layers[i]['dropout'].train()

    def eval(self):
        """Set to evaluation mode"""
        self.embedding_dropout.eval()
        self.classifier_dropout.eval()
        for i in range(self.num_layers):
            self.ffn_layers[i]['dropout'].eval()


# ============================================================================
# Training Function
# ============================================================================

def train_bert_classifier():
    """Train BERT classifier on sentiment data"""
    print("\n" + "=" * 60)
    print("Training BERT Classifier")
    print("=" * 60)

    # Configuration
    vocab_size = 30522
    hidden_size = 128  # Small for demo (BERT-base uses 768)
    num_heads = 4
    num_layers = 2     # Small for demo (BERT-base uses 12)
    num_classes = 2    # Binary classification
    max_seq_len = 32

    print("\nConfiguration:")
    print(f"  Vocab size: {vocab_size}")
    print(f"  Hidden size: {hidden_size}")
    print(f"  Attention heads: {num_heads}")
    print(f"  Encoder layers: {num_layers}")
    print(f"  Max sequence length: {max_seq_len}")

    # Create components
    tokenizer = SimpleTokenizer(vocab_size)
    dataset = SentimentDataset()
    model = BERTClassifier(vocab_size, hidden_size, num_heads, num_layers,
                           num_classes, max_seq_len)

    # Optimizer and loss
    params = model.parameters()
    optimizer = tz.optim.Adam(params, lr=0.001)
    criterion = tz.nn.CrossEntropyLoss()

    print(f"\nDataset size: {len(dataset)}")
    print(f"Parameters: {len(params)}")

    # Prepare all data
    all_input_ids = []
    all_labels = []

    for text, label in dataset.examples:
        token_ids = tokenizer.tokenize(text, max_seq_len)
        all_input_ids.append(token_ids)
        all_labels.append(label)

    input_ids_np = np.array(all_input_ids, dtype=np.int64)
    labels_np = np.array(all_labels, dtype=np.int64)

    # Training loop
    num_epochs = 10
    print(f"\nTraining for {num_epochs} epochs...")

    model.train()

    for epoch in range(num_epochs):
        optimizer.zero_grad()

        # Forward pass
        input_var = tz.Variable(tz.Tensor.from_numpy(input_ids_np), requires_grad=False)
        logits = model.forward(input_var)

        # Loss (CrossEntropyLoss expects Tensor for target, not Variable)
        labels_tensor = tz.Tensor.from_numpy(labels_np)
        loss = criterion(logits, labels_tensor)

        # Backward pass
        loss.backward()
        optimizer.step()

        # Calculate accuracy
        logits_np = logits.tensor().numpy()
        predictions = np.argmax(logits_np, axis=1)
        accuracy = np.mean(predictions == labels_np) * 100

        if (epoch + 1) % 2 == 0 or epoch == 0:
            print(f"Epoch {epoch+1:2d}/{num_epochs} | "
                  f"Loss: {loss.tensor().item():.4f} | "
                  f"Accuracy: {accuracy:.1f}%")

    # Evaluation
    print("\n" + "=" * 60)
    print("Evaluation")
    print("=" * 60)

    model.eval()

    with tz.no_grad():
        input_var = tz.Variable(tz.Tensor.from_numpy(input_ids_np), requires_grad=False)
        logits = model.forward(input_var)

        logits_np = logits.tensor().numpy()
        predictions = np.argmax(logits_np, axis=1)

        print("\nPredictions:")
        for i, (text, label) in enumerate(dataset.examples[:6]):
            pred = predictions[i]
            status = "CORRECT" if pred == label else "WRONG"
            sentiment = "Positive" if pred == 1 else "Negative"
            print(f"  [{status}] '{text[:40]}...' -> {sentiment}")

        final_accuracy = np.mean(predictions == labels_np) * 100
        print(f"\nFinal Accuracy: {final_accuracy:.1f}%")


# ============================================================================
# Component Demo
# ============================================================================

def demo_bert_components():
    """Demonstrate individual BERT components"""
    print("\n" + "=" * 60)
    print("BERT Components Demo")
    print("=" * 60)

    batch_size = 2
    seq_len = 16
    hidden_size = 64
    num_heads = 4

    # 1. Embedding layers
    print("\n[1] Embedding Layers")
    vocab_size = 1000

    token_emb = tz.nn.Embedding(vocab_size, hidden_size)
    pos_emb = tz.nn.Embedding(seq_len, hidden_size)

    # Create sample input
    input_ids = np.random.randint(0, vocab_size, (batch_size, seq_len)).astype(np.int64)
    pos_ids = np.tile(np.arange(seq_len), (batch_size, 1)).astype(np.int64)

    input_var = tz.Variable(tz.Tensor.from_numpy(input_ids), requires_grad=False)
    pos_var = tz.Variable(tz.Tensor.from_numpy(pos_ids), requires_grad=False)

    token_output = token_emb(input_var)
    pos_output = pos_emb(pos_var)

    print(f"  Token embedding: {input_ids.shape} -> {token_output.tensor().shape}")
    print(f"  Position embedding: {pos_ids.shape} -> {pos_output.tensor().shape}")

    # Combined embeddings
    combined = tz.Variable(token_output.tensor() + pos_output.tensor(), requires_grad=True)
    print(f"  Combined: {combined.tensor().shape}")

    # 2. Layer Normalization
    print("\n[2] Layer Normalization")
    layer_norm = tz.nn.LayerNorm([hidden_size])
    normed = layer_norm(combined)
    print(f"  Input: {combined.tensor().shape} -> Output: {normed.tensor().shape}")

    # 3. Multi-head Attention
    print("\n[3] Multi-head Self-Attention")
    attention = tz.nn.MultiheadAttention(hidden_size, num_heads)
    attn_out = attention(normed)  # Single input for self-attention
    print(f"  Input: {normed.tensor().shape}")
    print(f"  Attention output: {attn_out.tensor().shape}")

    # 4. GELU Activation
    print("\n[4] GELU Activation")
    test_input = np.array([-2, -1, 0, 1, 2], dtype=np.float32)
    test_var = tz.Variable(tz.Tensor.from_numpy(test_input), requires_grad=True)
    gelu_out = tz.nn.gelu(test_var)
    print(f"  Input: {list(test_input)}")
    print(f"  GELU:  {[f'{x:.3f}' for x in gelu_out.tensor().numpy()]}")

    # 5. Dropout
    print("\n[5] Dropout")
    dropout = tz.nn.Dropout(0.1)
    dropout.train()
    dropped = dropout(normed)
    print(f"  Dropout(0.1) applied to: {normed.tensor().shape}")

    print("\nAll BERT components demonstrated successfully!")


# ============================================================================
# Main
# ============================================================================

def main():
    tz.initialize()

    print("=" * 60)
    print("   BERT Text Classification Example")
    print("=" * 60)

    print("\nComponents demonstrated:")
    print("  Embeddings: Token, Position")
    print("  Layers: MultiheadAttention, LayerNorm, Linear")
    print("  Activations: GELU")
    print("  Training: CrossEntropyLoss, Adam")

    demo_bert_components()
    train_bert_classifier()

    print("\n" + "=" * 60)
    print("   BERT example completed successfully!")
    print("=" * 60)


if __name__ == "__main__":
    main()
